### ISSUE TITLE
[BUG] [CAPR] Upgrading Rancher during etcd restoration from < v2.7.5 to v2.7.5+ can lead to `implausible joined server for entry` 

### ISSUE THREAD

**Issue by Oats87** [rancher/rancher#42856]
**Rancher Server Setup**
- Rancher version: v2.7.6

**Information about the Cluster**
Downstream CAPR K3s or RKE2

**Describe the bug**
If a Rancher installation is upgraded from before v2.7.5 to v2.7.5 or newer during an etcd restoration operation, it is possible to have the planner start stalling out on `implausible joined server for entry`. 

**To Reproduce**
Run Rancher v2.7.3, and provision a cluster, break it, and attempt to restore it in a way that is not compatible with v2.7.3. Upgrade to v2.7.6 and attempt to restore the cluster -- observe that the planner will start to error out during etcd restoration shutdown plan generation as the `joined-to` annotation has not been set.

**Result**
Planner does not restore the cluster.

**Expected Result**
Planner restores the cluster.

**Screenshots**

**Additional context**

The temporary workaround for this issue is to annotate the machine plan secrets within the local cluster with the corresponding `joined-to` annotation that corresponds to the join URL of the new init node (the node you are going to restore the etcd snapshot to)

For RKE2 with an init node at `172.16.1.5` and a cluster called `my-rke2-cluster`, this would look like: 

```
kubectl annotate secret -n fleet-default -l cluster.x-k8s.io/cluster-name=my-rke2-cluster,rke.cattle.io/init-node=true rke.cattle.io/join-url=https://172.16.1.5:9345
kubectl annotate secret -n fleet-default -l cluster.x-k8s.io/cluster-name=my-rke2-cluster,rke.cattle.io/init-node!=true rke.cattle.io/joined-to=https://172.16.1.5:9345
```

For a K3s cluster with an init node at `172.16.1.5` and a cluster called `my-k3s-cluster`, this would look like:
```
kubectl annotate secret -n fleet-default -l cluster.x-k8s.io/cluster-name=my-rke2-cluster,rke.cattle.io/init-node=true rke.cattle.io/join-url=https://172.16.1.5:6443
kubectl annotate secret -n fleet-default -l cluster.x-k8s.io/cluster-name=my-rke2-cluster,rke.cattle.io/init-node!=true rke.cattle.io/joined-to=https://172.16.1.5:6443
```

Note the port difference between K3s and RKE2.

**Comment by Oats87** [comment]
Okay, I was able to come up with a solid reproduction to prove that the fix I am merging fixes this issue:

1. Provision 3 nodes total, and then provision a custom cluster with 1 controlplane+etcd node and 1 worker node using  K3s v1.23.17+k3s1 on Rancher v2.7.4 (NOTE: K3s version is important here because the newest K3s minors/patches have broken etcd snapshot functionality with 2.7.4)
2. Take 2 snapshots
3. Copy one of the snapshot files from the corresponding directory on the etcd node to your local machine
4. Go onto the server node and delete the snapshot from /var/lib/rancher/k3s/server/db/snapshots/<snapshot name>
5. Attempt to do an “etcd only” restore of the snapshot that you just deleted
6. Wait for the restore to fail “operation etcd restore failed”
7. Delete the etcd machine from the cluster
8. Upgrade to v2.7.9
9. Wait for the etcd machine to be deleted from the cluster
10. Add your second, clean machine to the cluster
11. Copy the etcd snapshot (NOT THE ONE YOU ARE ALREADY TRYING TO RESTORE) to the NEW controlplane+etcd machine, into `/var/lib/rancher/k3s/server/db/snapshots` (make the directory if you need to, i.e. `mkdir -p /var/lib/rancher/k3s/server/db/snapshots`)
12. Restore this snapshot by setting the `.spec.etcdSnapshotRestore.Name` of the `clusters.provisioning.cattle.io` object to the name of the file you copied to your new node
13. Observe that the cluster will start erroring with the `implausible joined server for entry` error
14. Upgrade the cluster to a version that contains my merged fix
15. Watch the cluster go through its etcd restoration and come back to life

**Comment by snasovich** [comment]
Moved out of v2.9-Next1 to v2.9-Next2 (with increased priority there) to free up QA capacity for higher priority issues.

**Comment by Josh-Diamond** [comment]
Ticket #42856 - Test Results - ✅ 

Verified on Rancher `v2.9-db81de3553c19f63d92666d13a71ba85a44e2142-head`:

| Scenario | Test Case | Result |
|---|---|---|
| 1. | Reproduce issue, then upgrade to `v2.9-head` and watch cluster self-resolve | ✅ |

---

Scenario 1 - ✅ 
1. Fresh install of HA Helm Rancher `v2.7.4`
2. Once `active`, provision a downstream Custom K3s cluster w/ k8s `v1.23.17+k3s1` - [ 1 etcd/cp + 1 wkr ]
3. Once `active`, take two snapshots
4. Move one to your local machine, and delete the other
5. Snap/restore to deleted snapshot w/ `etcd only` option selected
6. Observe `operation etcd restore failed` error message
7. Attempt to delete etcd machine from cluster - (mine just hung in same error state)
8. Upgrade Rancher to `v2.7.9`
9. Re-delete etcd machine from cluster - (this time it was successfully removed)
10. Register new `etcd/cp` node to custom cluster
11. Create snapshot directory on newly registered `etcd/cp` node by running the following command:
```log
mkdir -p /var/lib/rancher/k3s/server/db/snapshots
```
12. Move snapshot you're storing locally to directory you created in step 11, on newly registered `etcd/cp` node
13. Edit cluster object for downstream Custom k3s, and modify `.spec.etcdSnapshotRestore.Name` to reflect snapshot from step 12, then `save` yaml.  This should be formatted as shown below:
```yaml
spec:
  rkeConfig:
    etcdSnapshotRestore:
      name: <cluster_name>-<snapshot_name>-local
```
14. Reproduced
15. Upgrade Rancher to `v2.9-head`
16. Once upgraded, observe etcd restore proceed and cluster come up `active`
17. Verified - planner is able to restore the cluster; as expected

**Comment by martyav** [comment]
Noting that the fix will be release noted in the v2.9.2 draft
